cbits/src/mimalloc.rs

```diff
@@ -0,0 +1,289 @@
+//! static ALLOC: jj_cbits::mimalloc::MiMalloc = jj_cbits::mimalloc::MiMalloc;
+//! ```
+
+use core::alloc::GlobalAlloc;
+use core::ffi::{c_char, c_ulonglong, c_void, CStr};
+
+// Full set of external C APIs.
+//
+// NOTE: These should under no circumstances ever be made `pub`.
+extern "C" {
+    /// Allocate a block of memory of at least `size` bytes, aligned to the
+    /// given alignment, that is completely zeroed out.
+    fn mi_zalloc_aligned(size: usize, alignment: usize) -> *mut c_void;
+
+    /// Allocate a block of memory of at least `size` bytes, aligned to the
+    /// given alignment. The initial contents of the memory is unspecified, and
+    /// accessing it before initialization should be considered undefined
+    /// behavior.
+    fn mi_malloc_aligned(size: usize, alignment: usize) -> *mut c_void;
+
+    /// Reallocate a block of memory to a new size.
+    fn mi_realloc_aligned(p: *mut c_void, newsize: usize, alignment: usize) -> *mut c_void;
+
+    /// Free any previously allocated memory.
+    fn mi_free(p: *mut c_void);
+
+    /// Merge all thread-local statistics with the main statistics buffer, and
+    /// reset.
+    fn mi_stats_merge();
+
+    /// Reset the internal statistics counters.
+    fn mi_stats_reset();
+
+    /// Print global mimalloc statistics to a given output. The output is
+    /// currently fixed to `stderr`; both the first and second arguments must be
+    /// the NULL pointer.
+    fn mi_stats_print_out(out: unsafe extern "C" fn(*const c_char, *mut c_void), ctx: *mut c_void);
+
+    /// Register a "deferred free" function which will be called
+    /// deterministically over the course of the program.
+    fn mi_register_deferred_free(
+        f: unsafe extern "C" fn(bool, c_ulonglong, *mut c_void),
+        ctx: *mut c_void,
+    );
+}
+
+/// Provide global mimalloc heap statistics to a user-provided callback.
+///
+/// The provided callback function will be called multiple times in sequence,
+/// each time with a single argument, which is a single null-terminated line,
+/// represented as a [CStr]. Collectively, these lines will represent a summary
+/// of the global heap statistics for the entire program, meant to be written to
+/// the terminal or a log file.
+///
+/// Therefore, the simplest way to use this function is to simply provide a
+/// closure that prints the given log messages to `stderr` immediately:
+///
+/// ```rust,ignore
+/// eprintln!("========================================");
+/// eprintln!("mimalloc memory allocation statistics:\n");
+/// jj_cbits::mimalloc::stats_print(&|l| {
+///   eprint!("{}", l.to_string_lossy());
+/// });
+/// ```
+///
+/// Note that this merges all thread-local statistics into the main statistics
+/// summary before printing to `stderr`, so while it will give a global summary
+/// of the heap, it may cause some performance overhead while thread-local
+/// buffers are being flushed and merged.
+pub fn stats_print<F: Fn(&CStr)>(f: &'static F) {
```

Comment:
Should we convert the `CStr` eagerly here for the caller's benefit? I can't think of a reason someone might want otherwise.